File: src/vm.ts
import type { Renderer } from "./renderer";
import {
  CODON_MAP,
  type Codon,
  type CodonToken,
  Opcode,
  type VMState,
} from "./types";

/**
 * Stack values are 6-bit (0-63 range).
 * This is the maximum value used for normalization to coordinate/color spaces.
 */
const STACK_VALUE_RANGE = 64;

/**
 * Maximum degrees in a full circle for hue calculations.
 * Hue values from stack (0-63) are mapped to 0-360 degrees.
 */
const HUE_DEGREES = 360;

/**
 * Percentage scale (0-100) for lightness and saturation.
 * Stack values (0-63) are mapped to 0-100% range.
 */
const PERCENT_SCALE = 100;

/**
 * Scaling factor for general-purpose transformations.
 * Stack values are divided by this to produce reasonable scale multipliers.
 */
const SCALE_DIVISOR = 32;

/**
 * Default maximum instruction count to prevent infinite loops.
 * Programs exceeding this limit will throw an error.
 */
const DEFAULT_MAX_INSTRUCTIONS = 10000;

/**
 * Base-4 encoding weight for first digit position (4^2).
 * Used in decoding three-letter codons to 0-63 numeric literals.
 */
const BASE4_DIGIT1_WEIGHT = 16;

/**
 * Base-4 encoding weight for second digit position (4^1).
 * Used in decoding three-letter codons to 0-63 numeric literals.
 */
const BASE4_DIGIT2_WEIGHT = 4;

/**
 * Number of PUSH instructions used for LOOP parameters.
 * LOOP expects two values on stack (instructionCount, loopCount),
 * so we skip the last 2 history entries when replaying.
 */
const LOOP_PARAMETER_COUNT = 2;

/**
 * Boolean true value for comparison operations (EQ, LT).
 */
const BOOLEAN_TRUE = 1;

/**
 * Boolean false value for comparison operations (EQ, LT).
 */
const BOOLEAN_FALSE = 0;

/**
 * Virtual Machine interface for CodonCanvas execution.
 * Stack-based VM with drawing primitives and transform state.
 */
export interface VM {
  /** Current VM execution state (position, rotation, scale, color, stack) */
  state: VMState;
  /** Renderer for drawing operations */
  renderer: Renderer;

  /**
   * Execute a single opcode instruction.
   * @param opcode - The operation to execute
   * @param codon - The codon string (used for PUSH literal decoding)
   * @throws Error on stack underflow or invalid operations
   */
  execute(opcode: Opcode, codon: string): void;

  /**
   * Run entire genome program.
   * @param tokens - Array of codon tokens to execute
   * @returns Array of VM state snapshots (one per instruction) for timeline playback
   * @throws Error on instruction limit exceeded or execution errors
   */
  run(tokens: CodonToken[]): VMState[];

  /**
   * Reset VM to initial state.
   * Clears stack, resets position/rotation/scale/color, zeroes instruction counter.
   */
  reset(): void;

  /**
   * Create snapshot of current VM state.
   * @returns Deep copy of current state for rewind/step-through
   */
  snapshot(): VMState;

  /**
   * Restore VM from a previous snapshot.
   * @param state - Previously captured VM state
   */
  restore(state: VMState): void;
}

/**
 * CodonCanvas virtual machine implementation.
 * Executes DNA-like codon programs with stack-based graphics operations.
 *
 * Features:
 * - Stack machine with numeric values
 * - Drawing primitives (CIRCLE, RECT, LINE, TRIANGLE, ELLIPSE)
 * - Transform operations (TRANSLATE, ROTATE, SCALE, COLOR)
 * - Base-4 numeric literal encoding (0-63 range)
 * - Sandboxing with instruction limit (default: 10,000)
 * - State snapshot/restore for timeline scrubbing
 *
 * @example
 * ```typescript
 * const canvas = document.querySelector('canvas');
 * const renderer = new Canvas2DRenderer(canvas);
 * const vm = new CodonVM(renderer);
 *
 * const lexer = new CodonLexer();
 * const tokens = lexer.tokenize('ATG GAA CCC GGA TAA'); // PUSH 21, CIRCLE
 * const states = vm.run(tokens); // Execute and get state history
 * ```
 */
export class CodonVM implements VM {
  state: VMState;
  renderer: Renderer;
  private maxInstructions: number = DEFAULT_MAX_INSTRUCTIONS;
  private instructionHistory: {
    opcode: Opcode;
    codon: Codon;
    pushValue?: number;
  }[] = [];

  constructor(renderer: Renderer) {
    this.renderer = renderer;
    this.state = this.createInitialState();
  }

  private createInitialState(): VMState {
    return {
      position: { x: this.renderer.width / 2, y: this.renderer.height / 2 },
      rotation: 0,
      scale: 1,
      color: { h: 0, s: 0, l: 0 },
      stack: [],
      instructionPointer: 0,
      stateStack: [],
      instructionCount: 0,
      seed: Date.now(),
    };
  }

  reset(): void {
    this.state = this.createInitialState();
    this.renderer.clear();
    this.instructionHistory = [];
  }

  snapshot(): VMState {
    return JSON.parse(JSON.stringify(this.state));
  }

  restore(state: VMState): void {
    this.state = JSON.parse(JSON.stringify(state));
  }

  private pop(): number {
    const value = this.state.stack.pop();
    if (value === undefined) {
      throw new Error(
        `Stack underflow at instruction ${this.state.instructionPointer}`,
      );
    }
    return value;
  }

  private push(value: number): void {
    this.state.stack.push(value);
  }

  /**
   * Decode base-4 numeric literal from codon
   */
  private decodeNumericLiteral(codon: string): number {
    const baseMap: Record<string, number> = { A: 0, C: 1, G: 2, T: 3 };
    const d1 = baseMap[codon[0]] || 0;
    const d2 = baseMap[codon[1]] || 0;
    const d3 = baseMap[codon[2]] || 0;
    return d1 * BASE4_DIGIT1_WEIGHT + d2 * BASE4_DIGIT2_WEIGHT + d3;
  }

  /**
   * Scale value from 0-63 range to canvas coordinates
   */
  private scaleValue(value: number): number {
    return (value / STACK_VALUE_RANGE) * this.renderer.width;
  }

  execute(opcode: Opcode, codon: string): void {
    this.state.instructionCount++;

    if (this.state.instructionCount > this.maxInstructions) {
      throw new Error(
        `Instruction limit exceeded (max ${this.maxInstructions})`,
      );
    }

    // Track executed opcode for MIDI export
    this.state.lastOpcode = opcode;

    switch (opcode) {
      case Opcode.START:
        // Initialize execution
        break;

      case Opcode.STOP:
        // Terminate execution (handled by run loop)
        break;

      case Opcode.CIRCLE: {
        const radius = this.scaleValue(this.pop());
        this.renderer.circle(radius);
        break;
      }

      case Opcode.RECT: {
        const height = this.scaleValue(this.pop());
        const width = this.scaleValue(this.pop());
        this.renderer.rect(width, height);
        break;
      }

      case Opcode.LINE: {
        const length = this.scaleValue(this.pop());
        this.renderer.line(length);
        break;
      }

      case Opcode.TRIANGLE: {
        const size = this.scaleValue(this.pop());
        this.renderer.triangle(size);
        break;
      }

      case Opcode.ELLIPSE: {
        const ry = this.scaleValue(this.pop());
        const rx = this.scaleValue(this.pop());
        this.renderer.ellipse(rx, ry);
        break;
      }

      case Opcode.TRANSLATE: {
        const dy = this.scaleValue(this.pop());
        const dx = this.scaleValue(this.pop());
        this.renderer.translate(dx, dy);
        const transform = this.renderer.getCurrentTransform();
        this.state.position = { x: transform.x, y: transform.y };
        break;
      }

      case Opcode.ROTATE: {
        const degrees = this.pop();
        this.renderer.rotate(degrees);
        this.state.rotation = this.renderer.getCurrentTransform().rotation;
        break;
      }

      case Opcode.SCALE: {
        const factor = this.pop() / SCALE_DIVISOR;
        this.renderer.scale(factor);
        this.state.scale = this.renderer.getCurrentTransform().scale;
        break;
      }

      case Opcode.COLOR: {
        const l = this.pop() * (PERCENT_SCALE / STACK_VALUE_RANGE);
        const s = this.pop() * (PERCENT_SCALE / STACK_VALUE_RANGE);
        const h = this.pop() * (HUE_DEGREES / STACK_VALUE_RANGE);
        this.renderer.setColor(h, s, l);
        this.state.color = { h, s, l };
        break;
      }

      case Opcode.PUSH:
        // Push is handled in run() as it needs the next codon
        break;

      case Opcode.DUP: {
        const value = this.pop();
        this.push(value);
        this.push(value);
        break;
      }

      case Opcode.POP:
        this.pop();
        break;

      case Opcode.SWAP: {
        const b = this.pop();
        const a = this.pop();
        this.push(b);
        this.push(a);
        break;
      }

      case Opcode.NOP:
        // No operation
        break;

      case Opcode.SAVE_STATE: {
        const snapshot = this.snapshot();
        this.state.stateStack.push(snapshot);
        break;
      }

      case Opcode.RESTORE_STATE: {
        if (this.state.stateStack.length === 0) {
          throw new Error("RESTORE_STATE with empty state stack");
        }
        const savedState = this.state.stateStack.pop()!;
        // Restore transform state (position, rotation, scale, color)
        this.state.position = { ...savedState.position };
        this.state.rotation = savedState.rotation;
        this.state.scale = savedState.scale;
        this.state.color = { ...savedState.color };
        // Actually apply the restored transform to the renderer
        this.renderer.setPosition(savedState.position.x, savedState.position.y);
        this.renderer.setRotation(savedState.rotation);
        this.renderer.setScale(savedState.scale);
        this.renderer.setColor(
          savedState.color.h,
          savedState.color.s,
          savedState.color.l,
        );
        break;
      }

      case Opcode.ADD: {
        const b = this.pop();
        const a = this.pop();
        this.push(a + b);
        break;
      }

      case Opcode.MUL: {
        const b = this.pop();
        const a = this.pop();
        this.push(a * b);
        break;
      }

      case Opcode.SUB: {
        const b = this.pop();
        const a = this.pop();
        this.push(a - b);
        break;
      }

      case Opcode.DIV: {
        const b = this.pop();
        const a = this.pop();
        if (b === 0) {
          throw new Error(
            `Division by zero at instruction ${this.state.instructionPointer}`,
          );
        }
        this.push(Math.floor(a / b));
        break;
      }

      case Opcode.LOOP: {
        // Stack: [..., instructionCount, loopCount]
        // Replays the last N instructions M times
        const loopCount = this.pop();
        const instructionCount = this.pop();

        // Validate loop parameters
        if (loopCount < 0 || instructionCount < 0) {
          throw new Error(
            `LOOP requires non-negative parameters (count: ${loopCount}, instructions: ${instructionCount})`,
          );
        }

        if (instructionCount > this.instructionHistory.length) {
          throw new Error(
            `LOOP instruction count (${instructionCount}) exceeds history length (${this.instructionHistory.length})`,
          );
        }

        // Get the instructions to replay
        // Note: The last LOOP_PARAMETER_COUNT instructions in history are the PUSH operations for loop parameters
        // We want to replay instructions BEFORE those parameter PUSHes
        const historyBeforeParams =
          this.instructionHistory.length - LOOP_PARAMETER_COUNT;
        const startIdx = historyBeforeParams - instructionCount;
        const instructionsToRepeat = this.instructionHistory.slice(
          startIdx,
          historyBeforeParams,
        );

        // Execute the loop body loopCount times
        for (let iteration = 0; iteration < loopCount; iteration++) {
          for (const {
            opcode: loopOpcode,
            codon: loopCodon,
            pushValue,
          } of instructionsToRepeat) {
            // Check instruction limit
            this.state.instructionCount++;
            if (this.state.instructionCount > this.maxInstructions) {
              throw new Error(
                `Instruction limit exceeded (${this.maxInstructions})`,
              );
            }

            // Handle PUSH specially - use stored value instead of executing
            if (loopOpcode === Opcode.PUSH && pushValue !== undefined) {
              this.push(pushValue);
            } else {
              // Execute other instructions normally (don't add to history - avoid infinite growth)
              this.execute(loopOpcode, loopCodon);
            }
          }
        }
        break;
      }

      case Opcode.EQ: {
        // Equality comparison: [a, b] → [1 if a==b else 0]
        const b = this.pop();
        const a = this.pop();
        this.push(a === b ? BOOLEAN_TRUE : BOOLEAN_FALSE);
        break;
      }

      case Opcode.LT: {
        // Less than comparison: [a, b] → [1 if a<b else 0]
        const b = this.pop();
        const a = this.pop();
        this.push(a < b ? BOOLEAN_TRUE : BOOLEAN_FALSE);
        break;
      }

      default:
        throw new Error(`Unknown opcode: ${opcode}`);
    }
  }

  run(tokens: CodonToken[]): VMState[] {
    const snapshots: VMState[] = [];
    this.reset();

    let i = 0;
    while (i < tokens.length) {
      const token = tokens[i];
      const opcode = CODON_MAP[token.text];

      if (opcode === undefined) {
        throw new Error(
          `Unknown codon '${token.text}' at position ${token.position}`,
        );
      }

      this.state.instructionPointer = i;

      // Handle PUSH specially - reads next codon as numeric literal
      if (opcode === Opcode.PUSH) {
        i++;
        if (i >= tokens.length) {
          throw new Error(
            "PUSH instruction at end of program (missing numeric literal)",
          );
        }
        const literalCodon = tokens[i].text;
        const value = this.decodeNumericLiteral(literalCodon);
        this.push(value);
        this.instructionHistory.push({
          opcode: Opcode.PUSH,
          codon: token.text,
          pushValue: value,
        });
        snapshots.push(this.snapshot());
      } else if (opcode === Opcode.STOP) {
        // Stop execution
        snapshots.push(this.snapshot());
        break;
      } else {
        // Add non-LOOP instructions to history (LOOP doesn't get added to prevent recursion issues)
        if (opcode !== Opcode.LOOP) {
          this.instructionHistory.push({ opcode, codon: token.text });
        }
        this.execute(opcode, token.text);
        snapshots.push(this.snapshot());
      }

      i++;
    }

    return snapshots;
  }
}

File: engine/core/src/vm.rs
//! The stack-based CodonCanvas virtual machine.
//!
//! Unlike the original, which drew straight onto a `CanvasRenderingContext2D`,
//! this VM is renderer-agnostic: it emits an ordered list of [`DrawCommand`]s
//! that any backend (the Svelte canvas, an SVG exporter, a test harness) can
//! replay. Each command carries the full affine transform and color in effect
//! when it was issued, so a backend reproduces the original rendering with
//! `translate -> rotate -> scale -> draw-centered-at-origin`.
//!
//! # The negative-offset fix
//!
//! Stack values are 6-bit (`0..=63`). The original `scaleValue` mapped them to
//! `(v / 64) * width`, which is always `>= 0`, so `TRANSLATE` could only ever
//! move right and down — there was no way to express "move left" or "move up".
//! [`Vm::translate_offset`] fixes this by treating the value as *signed around
//! the midpoint 32*: `offset = ((v - 32) / 64) * width`. So `32` means "stay
//! put", values below 32 move left/up, and values above 32 move right/down.
//! Shape dimensions keep the original unsigned mapping ([`Vm::scale_value`]),
//! since a negative radius is meaningless.

use serde::Serialize;

use crate::codon::opcode_for;
use crate::lexer::Token;
use crate::opcode::Opcode;

const STACK_VALUE_RANGE: f64 = 64.0;
const HUE_DEGREES: f64 = 360.0;
const PERCENT_SCALE: f64 = 100.0;
const SCALE_DIVISOR: f64 = 32.0;
const DEFAULT_MAX_INSTRUCTIONS: u64 = 10_000;
const LOOP_PARAMETER_COUNT: usize = 2;
const MIN_SCALE: f64 = 0.001;
/// Stack value that maps to zero translation; see the module docs.
const TRANSLATE_CENTER: f64 = 32.0;

/// An HSL color: hue in degrees, saturation/lightness in percent.
#[derive(Debug, Clone, Copy, PartialEq, Serialize)]
pub struct Hsl {
    pub h: f64,
    pub s: f64,
    pub l: f64,
}

/// The affine transform in effect for a draw command.
#[derive(Debug, Clone, Copy, PartialEq, Serialize)]
pub struct Transform {
    pub x: f64,
    pub y: f64,
    /// Rotation in degrees, applied clockwise.
    pub rotation: f64,
    pub scale: f64,
}

/// A primitive shape, drawn centered at the transform origin.
#[derive(Debug, Clone, Copy, PartialEq, Serialize)]
#[serde(tag = "kind", rename_all = "lowercase")]
pub enum Shape {
    Circle { radius: f64 },
    Rect { width: f64, height: f64 },
    Line { length: f64 },
    Triangle { size: f64 },
    Ellipse { rx: f64, ry: f64 },
}

/// A single fill-and-stroke draw operation.
#[derive(Debug, Clone, Copy, PartialEq, Serialize)]
pub struct DrawCommand {
    pub transform: Transform,
    pub color: Hsl,
    pub shape: Shape,
}

/// A snapshot of VM state after one top-level instruction, for the timeline
/// scrubber and stack inspector.
#[derive(Debug, Clone, Serialize)]
pub struct Step {
    /// Token index of the instruction.
    pub index: usize,
    /// The codon executed.
    pub codon: String,
    /// Opcode display name.
    pub opcode: &'static str,
    /// Value pushed, for `PUSH` steps only.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub push_value: Option<f64>,
    /// Stack contents after the instruction (bottom-to-top).
    pub stack: Vec<f64>,
    pub transform: Transform,
    pub color: Hsl,
    /// Number of draw commands emitted so far; render `commands[..command_count]`
    /// to show the canvas at this step.
    pub command_count: usize,
    pub instruction_count: u64,
}

/// The result of running a genome.
#[derive(Debug, Clone, Serialize)]
pub struct RunResult {
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    pub width: f64,
    pub height: f64,
    pub commands: Vec<DrawCommand>,
    pub steps: Vec<Step>,
}

#[derive(Clone)]
struct HistItem {
    opcode: Opcode,
    codon: String,
    push_value: Option<f64>,
}

#[derive(Clone, Copy)]
struct SavedState {
    x: f64,
    y: f64,
    rotation: f64,
    scale: f64,
    color: Hsl,
}

/// The CodonCanvas virtual machine.
pub struct Vm {
    width: f64,
    height: f64,
    x: f64,
    y: f64,
    rotation: f64,
    scale: f64,
    color: Hsl,
    stack: Vec<f64>,
    state_stack: Vec<SavedState>,
    instruction_count: u64,
    max_instructions: u64,
    ip: usize,
    history: Vec<HistItem>,
    commands: Vec<DrawCommand>,
}

impl Vm {
    /// Creates a VM for a canvas of the given size, using the default
    /// instruction limit (10,000).
    pub fn new(width: f64, height: f64) -> Self {
        Self::with_limit(width, height, DEFAULT_MAX_INSTRUCTIONS)
    }

    /// Creates a VM with a custom instruction limit.
    pub fn with_limit(width: f64, height: f64, max_instructions: u64) -> Self {
        let mut vm = Self {
            width,
            height,
            x: 0.0,
            y: 0.0,
            rotation: 0.0,
            scale: 1.0,
            color: Hsl {
                h: 0.0,
                s: 0.0,
                l: 0.0,
            },
            stack: Vec::new(),
            state_stack: Vec::new(),
            instruction_count: 0,
            max_instructions: max_instructions.max(1),
            ip: 0,
            history: Vec::new(),
            commands: Vec::new(),
        };
        vm.reset();
        vm
    }

    fn reset(&mut self) {
        self.x = self.width / 2.0;
        self.y = self.height / 2.0;
        self.rotation = 0.0;
        self.scale = 1.0;
        self.color = Hsl {
            h: 0.0,
            s: 0.0,
            l: 0.0,
        };
        self.stack.clear();
        self.state_stack.clear();
        self.instruction_count = 0;
        self.ip = 0;
        self.history.clear();
        self.commands.clear();
    }

    fn current_transform(&self) -> Transform {
        Transform {
            x: self.x,
            y: self.y,
            rotation: self.rotation,
            scale: self.scale,
        }
    }

    /// Unsigned `0..=63 -> 0..width` mapping for shape dimensions.
    fn scale_value(&self, v: f64) -> f64 {
        (v / STACK_VALUE_RANGE) * self.width
    }

    /// Signed mapping for translation offsets; see the module docs for the fix.
    fn translate_offset(&self, v: f64) -> f64 {
        ((v - TRANSLATE_CENTER) / STACK_VALUE_RANGE) * self.width
    }

    fn pop(&mut self) -> Result<f64, String> {
        self.stack
            .pop()
            .ok_or_else(|| format!("Stack underflow at instruction {}", self.ip))
    }

    fn count_instruction(&mut self) -> Result<(), String> {
        self.instruction_count += 1;
        if self.instruction_count > self.max_instructions {
            return Err(format!(
                "Instruction limit exceeded (max {})",
                self.max_instructions
            ));
        }
        Ok(())
    }

    /// Runs a full token stream and returns the draw commands plus a per-step
    /// timeline. Hard errors stop execution but the partial render is retained.
    pub fn run(&mut self, tokens: &[Token]) -> RunResult {
        self.reset();
        let mut steps: Vec<Step> = Vec::new();
        let mut error: Option<String> = None;

        let mut i = 0;
        while i < tokens.len() {
            let token = &tokens[i];
            let Some(opcode) = opcode_for(&token.text) else {
                error = Some(format!(
                    "Unknown codon '{}' at position {}",
                    token.text, token.position
                ));
                break;
            };
            self.ip = i;
            let codon = token.text.clone();

            let step_result = match opcode {
                Opcode::Push => match self.execute_push(tokens, i) {
                    Ok((next, value)) => {
                        i = next;
                        Ok(Some(self.make_step(
                            self.ip,
                            codon.clone(),
                            opcode,
                            Some(value),
                        )))
                    }
                    Err(e) => Err(e),
                },
                Opcode::Stop => {
                    steps.push(self.make_step(i, codon, opcode, None));
                    break;
                }
                _ => {
                    if opcode != Opcode::Loop {
                        self.history.push(HistItem {
                            opcode,
                            codon: codon.clone(),
                            push_value: None,
                        });
                    }
                    self.execute(opcode, &codon)
                        .map(|()| Some(self.make_step(i, codon.clone(), opcode, None)))
                }
            };

            match step_result {
                Ok(Some(step)) => steps.push(step),
                Ok(None) => {}
                Err(e) => {
                    error = Some(e);
                    break;
                }
            }

            i += 1;
        }

        RunResult {
            ok: error.is_none(),
            error,
            width: self.width,
            height: self.height,
            commands: self.commands.clone(),
            steps,
        }
    }

    fn make_step(
        &self,
        index: usize,
        codon: String,
        opcode: Opcode,
        push_value: Option<f64>,
    ) -> Step {
        Step {
            index,
            codon,
            opcode: opcode.name(),
            push_value,
            stack: self.stack.clone(),
            transform: self.current_transform(),
            color: self.color,
            command_count: self.commands.len(),
            instruction_count: self.instruction_count,
        }
    }

    /// Handles `PUSH`: counts the instruction, reads the following codon as a
    /// base-4 literal, pushes it, and records it in history. Returns the index
    /// of the literal token (the caller advances past it) and the value.
    fn execute_push(&mut self, tokens: &[Token], i: usize) -> Result<(usize, f64), String> {
        self.count_instruction()?;
        let next = i + 1;
        if next >= tokens.len() {
            return Err("PUSH instruction at end of program (missing numeric literal)".into());
        }
        let value = f64::from(crate::base::decode_literal(&tokens[next].text));
        self.stack.push(value);
        self.history.push(HistItem {
            opcode: Opcode::Push,
            codon: tokens[i].text.clone(),
            push_value: Some(value),
        });
        Ok((next, value))
    }

    fn execute(&mut self, opcode: Opcode, codon: &str) -> Result<(), String> {
        self.count_instruction()?;
        match opcode {
            Opcode::Circle | Opcode::Rect | Opcode::Line | Opcode::Triangle | Opcode::Ellipse => {
                self.execute_draw(opcode)
            }
            Opcode::Translate | Opcode::Rotate | Opcode::Scale | Opcode::Color => {
                self.execute_transform(opcode)
            }
            Opcode::Add | Opcode::Sub | Opcode::Mul | Opcode::Div => self.execute_math(opcode),
            Opcode::Dup | Opcode::Pop | Opcode::Swap => self.execute_stack(opcode),
            Opcode::SaveState | Opcode::RestoreState => self.execute_state(opcode),
            Opcode::Loop => self.execute_loop(),
            Opcode::Eq | Opcode::Lt => self.execute_compare(opcode),
            Opcode::Start | Opcode::Stop | Opcode::Push | Opcode::Nop => Ok(()),
        }
        .map_err(|e: String| e)?;
        let _ = codon;
        Ok(())
    }

    fn emit(&mut self, shape: Shape) {
        self.commands.push(DrawCommand {
            transform: self.current_transform(),
            color: self.color,
            shape,
        });
    }

    fn execute_draw(&mut self, opcode: Opcode) -> Result<(), String> {
        match opcode {
            Opcode::Circle => {
                let v = self.pop()?;
                let radius = self.scale_value(v);
                self.emit(Shape::Circle { radius });
            }
            Opcode::Rect => {
                let h = self.pop()?;
                let w = self.pop()?;
                let height = self.scale_value(h);
                let width = self.scale_value(w);
                self.emit(Shape::Rect { width, height });
            }
            Opcode::Line => {
                let v = self.pop()?;
                let length = self.scale_value(v);
                self.emit(Shape::Line { length });
            }
            Opcode::Triangle => {
                let v = self.pop()?;
                let size = self.scale_value(v);
                self.emit(Shape::Triangle { size });
            }
            Opcode::Ellipse => {
                let ry_v = self.pop()?;
                let rx_v = self.pop()?;
                let ry = self.scale_value(ry_v);
                let rx = self.scale_value(rx_v);
                self.emit(Shape::Ellipse { rx, ry });
            }
            _ => unreachable!("execute_draw called with non-draw opcode"),
        }
        Ok(())
    }

    fn execute_transform(&mut self, opcode: Opcode) -> Result<(), String> {
        match opcode {
            Opcode::Translate => {
                // The negative-offset fix: signed offsets centered at 32.
                let dy_v = self.pop()?;
                let dx_v = self.pop()?;
                self.x += self.translate_offset(dx_v);
                self.y += self.translate_offset(dy_v);
            }
            Opcode::Rotate => {
                let degrees = self.pop()?;
                self.rotation += degrees;
            }
            Opcode::Scale => {
                let factor = self.pop()? / SCALE_DIVISOR;
                self.scale = (self.scale * factor).max(MIN_SCALE);
            }
            Opcode::Color => {
                let l = self.pop()? * (PERCENT_SCALE / STACK_VALUE_RANGE);
                let s = self.pop()? * (PERCENT_SCALE / STACK_VALUE_RANGE);
                let h = self.pop()? * (HUE_DEGREES / STACK_VALUE_RANGE);
                self.color = Hsl { h, s, l };
            }
            _ => unreachable!("execute_transform called with non-transform opcode"),
        }
        Ok(())
    }

    fn execute_math(&mut self, opcode: Opcode) -> Result<(), String> {
        let b = self.pop()?;
        let a = self.pop()?;
        let value = match opcode {
            Opcode::Add => a + b,
            Opcode::Mul => a * b,
            Opcode::Sub => a - b,
            Opcode::Div => {
                if b == 0.0 {
                    return Err(format!("Division by zero at instruction {}", self.ip));
                }
                (a / b).floor()
            }
            _ => unreachable!("execute_math called with non-arithmetic opcode"),
        };
        self.stack.push(value);
        Ok(())
    }

    fn execute_stack(&mut self, opcode: Opcode) -> Result<(), String> {
        match opcode {
            Opcode::Dup => {
                let v = self.pop()?;
                self.stack.push(v);
                self.stack.push(v);
            }
            Opcode::Pop => {
                self.pop()?;
            }
            Opcode::Swap => {
                let b = self.pop()?;
                let a = self.pop()?;
                self.stack.push(b);
                self.stack.push(a);
            }
            _ => unreachable!("execute_stack called with non-stack opcode"),
        }
        Ok(())
    }

    fn execute_state(&mut self, opcode: Opcode) -> Result<(), String> {
        match opcode {
            Opcode::SaveState => {
                self.state_stack.push(SavedState {
                    x: self.x,
                    y: self.y,
                    rotation: self.rotation,
                    scale: self.scale,
                    color: self.color,
                });
            }
            Opcode::RestoreState => {
                let saved = self
                    .state_stack
                    .pop()
                    .ok_or("RESTORE_STATE with empty state stack")?;
                self.x = saved.x;
                self.y = saved.y;
                self.rotation = saved.rotation;
                self.scale = saved.scale;
                self.color = saved.color;
            }
            _ => unreachable!("execute_state called with non-state opcode"),
        }
        Ok(())
    }

    // Float equality here is intentional: it mirrors the original `a === b`
    // semantics over integer-valued stacks.
    #[allow(clippy::float_cmp)]
    fn execute_compare(&mut self, opcode: Opcode) -> Result<(), String> {
        let b = self.pop()?;
        let a = self.pop()?;
        let result = match opcode {
            Opcode::Eq => f64::from(u8::from(a == b)),
            Opcode::Lt => f64::from(u8::from(a < b)),
            _ => unreachable!("execute_compare called with non-comparison opcode"),
        };
        self.stack.push(result);
        Ok(())
    }

    fn execute_loop(&mut self) -> Result<(), String> {
        let loop_count = self.pop()?;
        let instruction_count = self.pop()?;

        if loop_count < 0.0 || instruction_count < 0.0 {
            return Err(format!(
                "LOOP requires non-negative parameters (count: {loop_count}, instructions: {instruction_count})"
            ));
        }
        if self.history.len() < LOOP_PARAMETER_COUNT {
            return Err(format!(
                "LOOP requires at least {LOOP_PARAMETER_COUNT} instructions in history"
            ));
        }

        let max_replayable = self.history.len() - LOOP_PARAMETER_COUNT;
        let instruction_count = instruction_count as usize;
        if instruction_count > max_replayable {
            return Err(format!(
                "LOOP instruction count ({instruction_count}) exceeds available history ({max_replayable})"
            ));
        }

        let start = max_replayable - instruction_count;
        // Clone the slice so we can re-execute while mutating the VM.
        let body: Vec<HistItem> = self.history[start..max_replayable].to_vec();

        for _ in 0..(loop_count as u64) {
            for item in &body {
                self.replay(item)?;
            }
        }
        Ok(())
    }

    fn replay(&mut self, item: &HistItem) -> Result<(), String> {
        match (item.opcode, item.push_value) {
            (Opcode::Push, Some(value)) => {
                self.count_instruction()?;
                self.stack.push(value);
                Ok(())
            }
            _ => self.execute(item.opcode, &item.codon),
        }
    }
}

/// Convenience: tokenize and run a genome string in one call.
pub fn run_genome(source: &str, width: f64, height: f64) -> RunResult {
    match crate::lexer::tokenize(source) {
        Ok(tokens) => Vm::new(width, height).run(&tokens),
        Err(e) => RunResult {
            ok: false,
            error: Some(e),
            width,
            height,
            commands: Vec::new(),
            steps: Vec::new(),
        },
    }
}
